### PR TITLE
fix: revise blob.compose query parameters `if_generation_match`

### LINKED ISSUE #453: Change `blob.compose` to fit API behavior
Currently `blob.compose()` is misusing `if_generation_match` as a list of [`sourceObjects[].objectPreconditions.ifGenerationMatch`](https://cloud.google.com/storage/docs/json_api/v1/objects/compose#request-body) sent with the API request body. Based on the [doc](https://cloud.google.com/storage/docs/json_api/v1/objects/compose#parameters), `if_generation_match` is an optional query parameter that makes the operation conditional only when the composed object's current generation matches the given value.

### PR DESCRIPTION
**Pull request by cojenco**
This changes `if_generation_match`, `if_metageneration_match`, `if_source_generation_match` for  `blob.compose()` to fit API behavior

- Add pending deprecation warning in cases where `if_generation_match` and `if_metageneration_match` is passed in as a list of long. Add conditional handling for backwards compatibility.
-  Update `if_generation_match` to be correctly included within the API request query parameters if it is type long.  This makes the operation conditional on whether the **composed object's** current generation matches the given value. 
- Add `if_source_generation_match` as an optional parameter that serves as sourceObjects preconditions. This is sent with the API request body. When provided, the composition only performs if the generation of the **source object** that would be used matches this value.
- Revise docstring, example and tests

Fixes #453🦕

### CODE REVIEW

File: google/cloud/storage/blob.py
Comment (tseaver) [inline google/cloud/storage/blob.py:3049]: Given that `sources` is expected to be plural, this should likely expect that here, e.g.:

```(python)
            (Optional) Make the operation conditional on whether the current generation
            of each source blob matches the corresponding generation.
            The list must match ``sources`` item-to-item.
```

File: google/cloud/storage/blob.py
Comment (tseaver) [inline google/cloud/storage/blob.py:3081]: It would seem clearer to map the default value for `if_source_generation_match` as a sequence of `None` values, e.g.:

```python
        if if_source_generation_match is None:
            if_source_generation_match = [None] * sources_len
```
and then do the check:
```python
        if len(if_source_generation_match) != sources_len:
            raise ValueError(...)
```
I would also move both statements down to where they are being used below.

File: google/cloud/storage/blob.py
Comment (tseaver) [inline google/cloud/storage/blob.py:3101]: Given the change proposed above, and that `index` is used only to find the corresponding entry in `if_source_generation_match`, this would be clearer using the `zip` built-in, e.g.:

```python
        for source, source_generation in zip(sources, if_source_generation_match):
```

File: google/cloud/storage/blob.py
Comment (tseaver) [inline google/cloud/storage/blob.py:3105]: Clearer to just use the literal:

```python
            source_object = {"name": source.name, "generation": source.generation}
```

File: google/cloud/storage/blob.py
Comment (tseaver) [inline google/cloud/storage/blob.py:3112]: This would then all simplify down to:

```python
            if source_generation is not None:
                source_object["objectPreconditions"] = {"ifGenerationMatch": source_generation}
```

**Review by tseaver — CHANGES_REQUESTED** [review]

File: google/cloud/storage/blob.py
Comment (cojenco) [reply google/cloud/storage/blob.py]: Done, thanks @tseaver! This is much clearer. I like it a lot.

File: google/cloud/storage/blob.py
Comment (cojenco) [reply google/cloud/storage/blob.py:3127]: Thanks @tseaver! That makes sense. I've made the changes and grouped them closer together. Though wouldn't we want to do the check and raise the exception early in the method? WDYT?

Comment (tseaver) [reply google/cloud/storage/blob.py:3127]: @cojenco 
> Wouldn't we want to do the check and raise the exception early in the method?

I like the clarity of doing all the manipulations which depend on `if_source_generation_match` in one place:  in the odd case where the user passes an invalid sequence, we will have done only a trivial amount of work populating `query_params` before raising.

**Review by tseaver — APPROVED** [review]

**Comment by tseaver** [comment]
@cojenco I will merge / repair the tests broken by my PRs this week.

**Comment by tseaver** [comment]
@googlebot I consent.

**Comment by tseaver** [comment]
@cojenco I put back the `do not merge` label to allow us to consider whether this might be a breaking change.

**Comment by cojenco** [comment]
Got it, thanks @tseaver!

**Comment by cojenco** [comment]
This currently will introduce a breaking change. Revising PR for backwards compatibility.

**Comment by cojenco** [comment]
Updated PR to handle backwards compatibility. Ready for review:)

File: google/cloud/storage/blob.py
Comment (tseaver) [inline google/cloud/storage/blob.py:2989]: This can be a regular `DeprecationWarning` -- we are already deprecating the usage, not "fixin' to commence" to deprecated it.  Also, we want to use `stacklevel=2` to highlight the call site.

File: google/cloud/storage/blob.py
Comment (tseaver) [inline google/cloud/storage/blob.py:2946]: Ouch!  PR #464 is also mangling the docstrings for the conditional parameters.

File: google/cloud/storage/blob.py
Comment (tseaver) [inline google/cloud/storage/blob.py:3007]: Likewise here.

File: tests/system/test_system.py
Comment (tseaver) [inline tests/system/test_system.py]: Do we need a system test for the deprecated case?

**Review by tseaver — CHANGES_REQUESTED** [review]

File: google/cloud/storage/blob.py
Comment (cojenco) [reply google/cloud/storage/blob.py]: Done. Also revised the warning message based on #464. Thanks for catching this.

File: tests/system/test_system.py
Comment (cojenco) [inline tests/system/test_system.py:1752]: @tseaver Re: Do we need a system test for the deprecated case?
`test_compose_with_generation_match_list` tests both deprecated cases that would fail or succeed. Here we test operation to succeed with a deprecated case in which list of long is passed in and source objects generations match. Are there any other cases missing?

File: google/cloud/storage/blob.py
Comment (cojenco) [reply google/cloud/storage/blob.py:2946]: Thanks for catching this. I think this is a special case where we will need to have a full explanation in the docstrings considering deprecation and backwards compatibility. Let me know if there is a better way to handle the docstrings here.

File: tests/system/test_system.py
Comment (tseaver) [reply tests/system/test_system.py:1752]: I was actually questioning the need for systests for the deprecated case(s), rather than arguing for more. :)

File: google/cloud/storage/blob.py
Comment (tseaver) [reply google/cloud/storage/blob.py:2946]: I think I'll hold off merging #464 until this lands, and then look at how to merge this change.

File: google/cloud/storage/blob.py
Comment (andrewsg) [inline google/cloud/storage/blob.py:2942]: I would recommend phrasing this something like:

"Note: In a previous version, this argument worked identically to the if_source_generation_match argument. For backwards-compatibility reasons, if a list is passed in, this argument will behave like if_source_generation_match and also issue a DeprecationWarning."

File: google/cloud/storage/blob.py
Comment (andrewsg) [inline google/cloud/storage/blob.py:2954]: This didn't behave like if_source_metageneration_match before? It just did nothing?

File: google/cloud/storage/blob.py
Comment (andrewsg) [inline google/cloud/storage/blob.py:2987]: This warning text sounds good to me.

File: google/cloud/storage/blob.py
Comment (andrewsg) [inline google/cloud/storage/blob.py:2995]: I'd recommend the first line read "Use if_generation_match to match the generation of the destination object by passing in a generation number, instead of a list."

**Review by andrewsg — CHANGES_REQUESTED** [review]

File: google/cloud/storage/blob.py
Comment (cojenco) [reply google/cloud/storage/blob.py:2954]: I don't think source objects metagenerations checks are performed here, according to the [API doc](https://cloud.google.com/storage/docs/json_api/v1/objects/compose#request-body).

File: google/cloud/storage/blob.py
Comment (cojenco) [reply google/cloud/storage/blob.py]: Done. I like this phrasing a lot, thanks!

**Review by tseaver — APPROVED** [review]

**Review by andrewsg — APPROVED** [review]
Thanks!

**Pull request merged**
